labelbox/schema/dataset.py

```diff
@@ -749,3 +765,94 @@ def _export(
         res = res[mutation_name]
         task_id = res["taskId"]
         return Task.get_task(self.client, task_id)
+
+    def upsert_data_rows(self, specs: List[DataRowSpec]) -> "Task":
+        """
+        Upserts data rows in this dataset.
+
+        >>>     task = dataset.upsert_data_rows([
+        >>>         # create new data row
+        >>>         DataRowSpec(
+        >>>             row_data="http://my_site.com/photos/img_01.jpg",
+        >>>             global_key="global_key1",
+        >>>             external_id="ex_id1",
+        >>>             attachments=[
+        >>>                 DataRowAttachmentSpec(type=AttachmentType.RAW_TEXT, name="att1", value="test1")
+        >>>             ],
+        >>>             metadata=[
+        >>>                 DataRowMetadataSpec(name="tag", value="tag value"),
+        >>>             ]
+        >>>         ),
+        >>>         # update existing data row by global key
+        >>>         DataRowSpec(
+        >>>             global_key="global_key1",
+        >>>             external_id="ex_id1_updated"
+        >>>         ),
+        >>>         # update global key of data row by existing global key
+        >>>         DataRowSpec(
+        >>>             key=DataRowGlobalKey("global_key1"),
+        >>>             global_key="global_key1_updated"
+        >>>         ),
+        >>>         # update data row by ID
+        >>>         DataRowSpec(
+        >>>             key=DataRowIdKey(dr.uid),
+        >>>             external_id="ex_id1_updated"
+        >>>         ),
+        >>>     ])
+        >>>     task.wait_till_done()
+        """
+        if len(specs) > MAX_DATAROW_PER_API_OPERATION:
+            raise MalformedQueryException(
+                f"Cannot upsert more than {MAX_DATAROW_PER_API_OPERATION} DataRows per function call."
+            )
+
+        class ManifestFile:
```

Comment:
I understand functions within functions is a pattern already used in this file but I would avoid this pattern unless there's not viable alternative - there's a performance penalty, not to mention it makes the code more difficult to read.

Reply:
I didn't want to use a class in the first place. Sync up with @vbrodsky

Reply:
My comment was not so much in the class as the functions within functions, personally I would extract those to avoid the performance penalty as I do not think you need the closure.

Reply:
I agree solely on readability concerns, no need to nest the function instead of making it a private func if it doesn't use anything from the outer scope.

Or even just not having a function for this at alll; the function is defined inline right before it's used for the only time so it's not like it's encapsulating anything or reducing line count.